Getting the strength for all edges of a network

Manas-7854 opened the pull request:
Issue #1684 
- Added edge_strength function inside DAG.py
- added test to test the function

Manas-7854 committed:
edge strength for DAG

Manas-7854 committed:
Merge branch 'pgmpy:dev' into dev

Manas-7854 committed:
fixed formatting

Manas-7854 committed:
Merge remote-tracking branch 'refs/remotes/origin/dev' into dev

Nimish-4 commented:
@Manas-7854 The tests are failing due to formatting issues. pgmpy uses `black`, `isort` and a few other formatting checks. You can find them in the ".pre-commit-config.yaml" file in the codebase. Install them locally on your editor, use them to format the code and then the formatting checks will pass.

Manas-7854 committed:
fixed formatting

Manas-7854 committed:
fixed parameter issue on calling pillai_trace

Nimish-4 commented on .gitignore:
```diff
@@ -19,6 +19,7 @@ develop-eggs
 .installed.cfg
 lib
 lib64
+myenv
```
This looks related to your local environment. If yes, I think it doesn't need to be here and should be removed from the changes.

Manas-7854 replied:
is there a way i can use `git add .` and the myenv folder doesn't get commited or do i need to add all the files i have changed manually?

Nimish-4 replied:
@Manas-7854 You could look it up, but I don't think it can be done like that. I prefer to do it manually, also helps in keeping track of what is being committed.

jgyasu replied:
> is there a way i can use git add . and the myenv folder doesn't get commited or do i need to add all the files i have changed manually?

you can also create a `.gitignore` file and add the files and folders there that you don't want to commit

avinab-neogy replied:
The `myenv` directory is already included in the `.gitignore` file, so it shouldn't be an issue right? @Nimish-4

Nimish-4 replied:
@avinabneogy23 It can be present in your local project, but it doesn't need to be pushed to the pgmpy repo is what I'm saying.

avinab-neogy replied:
I see, thanks @Nimish-4

ankurankan replied:
@avinabneogy23 In general, it is always a better idea to name each file (as @Nimish-4 also mentioned) when doing a git add: `git add <filename1> <filename2> ...`. This makes sure that you are adding the correct changes. Additionally, you can always run `git diff --cached` after doing add so that you can double check which changes are going to get committed.

Nimish-4 commented on model.bif:
```diff
@@ -0,0 +1,31 @@
+probability ( D | C ) {
+    ( 0 ) 0.11202275075549158, 0.46515371328405486, 0.08913003410201674, 0.33369350185843677;
+
+}
```
This model file is created by a test that does not clean up the files it created. Should also be removed.

Nimish-4 commented on pgmpy/base/DAG.py:
```diff
@@ -1257,6 +1257,82 @@ def copy(self):
+        >>> dag.edge_strength(data, edges=('X', 'Y'))
+        {('X', 'Y'): 1.2260921400386593e-07}
+        """
+        from pgmpy.estimators.CITests import pillai_trace
```
Better to import it at the top with all the other imports.

Manas-7854 replied:
Hi, importing at the top is giving the following error `test_DAG.py - TypeError: module() takes at most 2 arguments (3 given)`.
Also there are other functions that import inside the function like for example `fit`.

Nimish-4 replied:
@Manas-7854 Keep your original implementation, in that case!

ankurankan replied:
Importing outside might be resulting in a circular import.

Nimish-4 commented on pgmpy/base/DAG.py:
```diff
@@ -1257,6 +1257,82 @@ def copy(self):
+            conditioning_set = set(pa_X + pa_Y) - {x, y}
+
+            # Run CI test and get effect size
+            result = pillai_trace(
```
If any other tests are suitable, or were discussed previously, you can create a new argument for the user to choose a CI test.

Manas-7854 replied:
> Solution: Add a method to DAG class to get the edge strength for every edge in the model. The edge strength can be computed using the effect size measure of ci_pillai test. For an edge X -> Y, the edge strength would be the effect size of the test `X _|_ Y | pa(X) U pa(Y)`.

in the issue only pilliai_test was discussed, however i can allow the user to select the test if that is required?

Nimish-4 replied:
In that case, it's fine! @ankurankan will let us know if anything else is required.

ankurankan replied:
Yeah, I would keep it just as pillai trace for now as it can work with all types of data. We also do not have an effect size measure for all the tests. In future, we can expand this.

Nimish-4 commented on pgmpy/tests/test_base/test_DAG.py:
```diff
@@ -360,6 +360,39 @@ def test_dag_fit(self):
+        empty_dag = DAG()
+        data = pd.DataFrame({"a": [0, 1, 0, 1], "b": [1, 3, 0, 2], "c": [1, 1, 0, 0]})
+        self.assertEqual(empty_dag.edge_strength(data), {})
+
```
I think adding a test where the output of the function is tested would be great. In other words, some numerical comparison of the output and expected output.

ankurankan replied:
I agree, we can use a LinearGaussian BN, simulate data from it and try to estimate the edge strengths. The edge strength should match the alpha parameter of the LinearGaussian CPD

ankurankan replied:
We should also add some tests that cover the latent variable case.

Manas-7854 replied:
hi I am having a little hard time understanding how to generate data from the linear gaussian, can you help me with that?
here is what i have tried to implement but its not working since the edge_strengths from `DAG.edge_strength` do not match with alpha parameter. 

```  
def test_edge_strength_basic(self):
        dag = DAG([("X", "Y"), ("Z", "Y")])

        num_samples = 1000
        seed=42

        np.random.seed(seed)

        # Root nodes: X and Z
        X = np.random.normal(loc=0, scale=1, size=num_samples)
        Z = np.random.normal(loc=0, scale=1, size=num_samples)

        # Child node: Y = 0.9 + (-2.0 * X) + (1.5 * Z) + noise
        noise = np.random.normal(loc=0, scale=1.0, size=num_samples)
        Y = 0.9 + (-2.0 * X) + (1.5 * Z) + noise

        data= pd.DataFrame({"X": X, "Z": Z, "Y": Y})

        # Define true CPD with known beta
        cpd = LinearGaussianCPD("Y", beta=[0.9, -2.0, 1.5], std=1.0, evidence=["X", "Z"])

        # Ground truth edge strengths from beta (skip intercept)
        true_strengths = {}
        for i, parent in enumerate(cpd.evidence):
            true_strengths[(parent, cpd.variable)] = cpd.beta[i + 1]  # skip intercept at index 0


        # Estimated edge strengths from data
        estimated_strengths = dag.edge_strength(data=data)

        tolerance = 0.1
        for edge, true_val in true_strengths.items():
            est_val = estimated_strengths[edge]
            assert abs(abs(est_val) - abs(true_val)) <= tolerance, f"Edge {edge}: expected ~{true_val}, got {est_val}"
```

ankurankan replied:
I think I have found an issue with `pillai_trace`, so the values wouldn't match for now (#2066). You can write the test with the values that you are getting now and we will fix the tests after #2066 is fixed.

ankurankan replied:
You can simulate data from a linear model as:
```python
In [47]: from pgmpy.estimators.CITests import pearsonr, pillai_trace

In [48]: from pgmpy.models import LinearGaussianBayesianNetwork as LGBN

In [49]: model = get_example_model("alarm")

In [50]: linear_model = LGBN([("X", "Y"), ("Z", "Y")])

In [51]: x_cpd = LinearGaussianCPD(variable='X', beta=[0], std=1)

In [52]: y_cpd = LinearGaussianCPD(variable='Y', beta=[0, 0.4, 0.6], std=1, evidence=['X', 'Z'])

In [53]: z_cpd = LinearGaussianCPD(variable='Z', beta=[0], std=1)

In [54]: linear_model.add_cpds(x_cpd, y_cpd, z_cpd)

In [55]: df = linear_model.simulate(n_samples=int(1e4))

```

Manas-7854 replied:
just to clarify, should i keep the tests as i had initially implemented?

Manas-7854 committed:
updated test to check for numerical outputs

Manas-7854 committed:
removed model.bif and updated gitignore for myenv

ankurankan commented on pgmpy/base/DAG.py:
```diff
@@ -1257,6 +1257,82 @@ def copy(self):
 
+    def edge_strength(self, data, edges=None):
+        """
+        Computes the strength of edges in the DAG using conditional independence tests.
```
This will need to be expanded by a lot. Pillai's trace is not a standard method (the research paper still hasn't been published), so we will need to give the users a better idea of what numbers they are seeing. The following things need to be mentioned in the documentation:

1. Uses a residualization based approach (similar to partial correlation) and uses the XGBoost estimator instead of linear regression.
2. Computes the Pillai Trace effect size based on canonical correlations instead of using correlation (as is the case with partial correlation)
3. Measure of linear relationship between the residuals.
1. Works for any mixture of categorical and continuous variables.
2. The value is bounded between 0 and 1.

Interpretation:

1. If the edge is between two continuous variables, the strength would be equivalent to partial correlation.
4. If the edge is between two categorical variables, the strength would be analogous to Cramers' V.

ankurankan commented on pgmpy/base/DAG.py:
```diff
@@ -1257,6 +1257,82 @@ def copy(self):
+
+        Examples
+        --------
+        >>> import pandas as pd
```
Please use a Linear Gaussian BN for the Example here.

Manas-7854 replied:
do i need to use Linear Gaussian BN for tests as well?

ankurankan commented on pgmpy/base/DAG.py:
```diff
@@ -1257,6 +1257,82 @@ def copy(self):
+            x, y = edge
+
+            # Check if either x or y is a latent node
+            if x in self.latents or y in self.latents:
```
The check needs to be:
1. Fail if either X or Y are latents
2. Fail if any of the parents of X and Y are latents.

ankurankan commented on pgmpy/base/DAG.py:
```diff
@@ -1257,6 +1257,82 @@ def copy(self):
+                )
+
+            # Get parents of x and y
+            pa_X = list(self.predecessors(x))
```
Use the `DAG.get_parents` method here. The definition of predecessors in case of causal inference and graph theory are different. In causal inference predecessor means: parents + parents of parents + their parents ... whereas in standard graph theory predecessors are defined as just the parents. 

The get_parents method makes a call to the predecessors as well, but using `get_parents` makes it more explicit here.

ankurankan commented on pgmpy/base/DAG.py:
```diff
@@ -1257,6 +1257,82 @@ def copy(self):
+            )
+
+            # Store the edge strength
+            strengths[edge] = result[1]
```
I think result is (effect_size, p_value), and the effect size represents the strength, so this should be `result[0]`.

ankurankan reviewed (changes requested):
@Manas-7854 Thanks for the PR. I think most of the code structure is here now, but the code and some logic still needs work.

Manas-7854 committed:
updated docsting, used dag.get_parents, fixed index for result,

ankurankan commented on pgmpy/base/DAG.py:
```diff
@@ -1257,6 +1257,82 @@ def copy(self):
+            pa_Y = list(self.predecessors(y))
+
+            # Combine parents for conditioning set (excluding x and y themselves)
+            conditioning_set = set(pa_X + pa_Y) - {x, y}
```
@Manas-7854 Sorry, slight change. This should be only be `pa_Y - {x, y}`

Manas-7854 committed:
updated conditioning set

Manas-7854 committed:
Merge branch 'pgmpy:dev' into dev

Manas-7854 committed:
Merge remote-tracking branch 'refs/remotes/origin/dev' into dev

Manas-7854 committed:
import os
import unittest
import warnings

import networkx as nx
import numpy as np
import pandas as pd

import pgmpy.tests.help_functions as hf
from pgmpy.base import DAG, PDAG
from pgmpy.estimators import (
    BayesianEstimator,
    ExpectationMaximization,
    MaximumLikelihoodEstimator,
)
from pgmpy.factors.continuous import LinearGaussianCPD
from pgmpy.factors.discrete import TabularCPD
from pgmpy.models import DiscreteBayesianNetwork

class TestDAGCreation(unittest.TestCase):
    def setUp(self):
        self.graph = DAG()

    def test_class_init_without_data(self):
        self.assertIsInstance(self.graph, DAG)

    def test_class_init_with_data_string(self):
        self.graph = DAG([("a", "b"), ("b", "c")])
        self.assertListEqual(sorted(self.graph.nodes()), ["a", "b", "c"])
        self.assertListEqual(
            hf.recursive_sorted(self.graph.edges()), [["a", "b"], ["b", "c"]]
        )
        self.assertEqual(self.graph.latents, set())

        self.graph = DAG([("a", "b"), ("b", "c")], latents=["b"])
        self.assertListEqual(sorted(self.graph.nodes()), ["a", "b", "c"])
        self.assertListEqual(
            hf.recursive_sorted(self.graph.edges()), [["a", "b"], ["b", "c"]]
        )
        self.assertEqual(self.graph.latents, set(["b"]))

    def test_class_init_with_adj_matrix_dict_of_dict(self):
        adj = {"a": {"b": 4, "c": 3}, "b": {"c": 2}}
        self.graph = DAG(adj, latents=set(["a"]))
        self.assertEqual(self.graph.latents, set("a"))
        self.assertListEqual(sorted(self.graph.nodes()), ["a", "b", "c"])
        self.assertEqual(self.graph.adj["a"]["c"]["weight"], 3)

    def test_class_init_with_adj_matrix_dict_of_list(self):
        adj = {"a": ["b", "c"], "b": ["c"]}
        self.graph = DAG(adj, latents=set(["a"]))
        self.assertEqual(self.graph.latents, set("a"))
        self.assertListEqual(sorted(self.graph.nodes()), ["a", "b", "c"])

    def test_class_init_with_pd_adj_df(self):
        df = pd.DataFrame([[0, 3], [0, 0]])
        self.graph = DAG(df, latents=set([0]))
        self.assertEqual(self.graph.latents, set([0]))
        self.assertListEqual(sorted(self.graph.nodes()), [0, 1])
        self.assertEqual(self.graph.adj[0][1]["weight"], {"weight": 3})  # None

    def test_variable_name_contains_non_string_adj_matrix(self):
        df = pd.DataFrame([[0, 3], [0, 0]])
        self.graph = DAG(df)
        self.assertEqual(self.graph._variable_name_contains_non_string(), (0, int))

    def test_variable_name_contains_non_string_mixed_types(self):
        self.graph = DAG([("a", "b"), ("b", "c"), ("a", 3.2)])
        self.graph.nodes()
        self.assertEqual(self.graph._variable_name_contains_non_string(), (3.2, float))

    def test_add_node_string(self):
        self.graph = DAG()
        self.graph.add_node("a")
        self.assertListEqual(list(self.graph.nodes()), ["a"])
        self.assertEqual(self.graph.latents, set())

        self.graph = DAG()
        self.graph.add_node("a", latent=True)
        self.assertListEqual(list(self.graph.nodes()), ["a"])
        self.assertEqual(self.graph.latents, set(["a"]))

    def test_add_node_nonstring(self):
        self.graph = DAG()
        self.graph.add_node(1)

        self.graph = DAG()
        self.graph.add_node(1, latent=True)

    def test_add_nodes_from_string(self):
        self.graph = DAG()
        self.graph.add_nodes_from(["a", "b", "c", "d"])
        self.assertListEqual(sorted(self.graph.nodes()), ["a", "b", "c", "d"])
        self.assertEqual(self.graph.latents, set())

        self.graph = DAG()
        self.graph.add_nodes_from(["a", "b", "c", "d"], latent=True)
        self.assertListEqual(sorted(self.graph.nodes()), ["a", "b", "c", "d"])
        self.assertEqual(self.graph.latents, set(["a", "b", "c", "d"]))

        self.graph = DAG()
        self.graph.add_nodes_from(
            ["a", "b", "c", "d"], latent=[True, False, True, False]
        )
        self.assertListEqual(sorted(self.graph.nodes()), ["a", "b", "c", "d"])
        self.assertEqual(self.graph.latents, set(["a", "c"]))

    def test_add_nodes_from_non_string(self):
        self.graph = DAG()
        self.graph.add_nodes_from([1, 2, 3, 4])

        self.graph = DAG()
        self.graph.add_nodes_from([1, 2, 3, 4], latent=True)

        self.graph = DAG()
        self.graph.add_nodes_from([1, 2, 3, 4], latent=[True, False, False, False])

    def test_add_node_weight(self):
        self.graph = DAG()
        self.graph.add_node("weighted_a", weight=0.3)
        self.assertEqual(self.graph.nodes["weighted_a"]["weight"], 0.3)
        self.assertEqual(self.graph.latents, set())

        self.graph = DAG()
        self.graph.add_node("weighted_a", weight=0.3, latent=True)
        self.assertEqual(self.graph.nodes["weighted_a"]["weight"], 0.3)
        self.assertEqual(self.graph.latents, set(["weighted_a"]))

    def test_add_nodes_from_weight(self):
        self.graph = DAG()
        self.graph.add_nodes_from(["weighted_b", "weighted_c"], weights=[0.5, 0.6])
        self.assertEqual(self.graph.nodes["weighted_b"]["weight"], 0.5)
        self.assertEqual(self.graph.nodes["weighted_c"]["weight"], 0.6)

        self.graph = DAG()
        self.graph.add_nodes_from(
            ["weighted_b", "weighted_c"], weights=[0.5, 0.6], latent=True
        )
        self.assertEqual(self.graph.nodes["weighted_b"]["weight"], 0.5)
        self.assertEqual(self.graph.nodes["weighted_c"]["weight"], 0.6)
        self.assertEqual(self.graph.latents, set(["weighted_b", "weighted_c"]))

        self.graph = DAG()
        self.graph.add_nodes_from(
            ["weighted_b", "weighted_c"], weights=[0.5, 0.6], latent=[True, False]
        )
        self.assertEqual(self.graph.nodes["weighted_b"]["weight"], 0.5)
        self.assertEqual(self.graph.nodes["weighted_c"]["weight"], 0.6)
        self.assertEqual(self.graph.latents, set(["weighted_b"]))

        self.graph.add_nodes_from(["e", "f"])
        self.assertEqual(self.graph.nodes["e"]["weight"], None)
        self.assertEqual(self.graph.nodes["f"]["weight"], None)

    def test_add_edge_string(self):
        self.graph.add_edge("d", "e")
        self.assertListEqual(sorted(self.graph.nodes()), ["d", "e"])
        self.assertListEqual(list(self.graph.edges()), [("d", "e")])
        self.graph.add_nodes_from(["a", "b", "c"])
        self.graph.add_edge("a", "b")
        self.assertListEqual(
            hf.recursive_sorted(self.graph.edges()), [["a", "b"], ["d", "e"]]
        )

    def test_add_edge_nonstring(self):
        self.graph.add_edge(1, 2)

    def test_add_edges_from_string(self):
        self.graph.add_edges_from([("a", "b"), ("b", "c")])
        self.assertListEqual(sorted(self.graph.nodes()), ["a", "b", "c"])
        self.assertListEqual(
            hf.recursive_sorted(self.graph.edges()), [["a", "b"], ["b", "c"]]
        )
        self.graph.add_nodes_from(["d", "e", "f"])
        self.graph.add_edges_from([("d", "e"), ("e", "f")])
        self.assertListEqual(sorted(self.graph.nodes()), ["a", "b", "c", "d", "e", "f"])
        self.assertListEqual(
            hf.recursive_sorted(self.graph.edges()),
            hf.recursive_sorted([("a", "b"), ("b", "c"), ("d", "e"), ("e", "f")]),
        )

    def test_add_edges_from_nonstring(self):
        self.graph.add_edges_from([(1, 2), (2, 3)])

    def test_add_edge_weight(self):
        self.graph.add_edge("a", "b", weight=0.3)
        self.assertEqual(self.graph.adj["a"]["b"]["weight"], 0.3)

    def test_add_edges_from_weight(self):
        self.graph.add_edges_from([("b", "c"), ("c", "d")], weights=[0.5, 0.6])
        self.assertEqual(self.graph.adj["b"]["c"]["weight"], 0.5)
        self.assertEqual(self.graph.adj["c"]["d"]["weight"], 0.6)

        self.graph.add_edges_from([("e", "f")])
        self.assertEqual(self.graph.adj["e"]["f"]["weight"], None)

    def test_update_node_parents_bm_constructor(self):
        self.graph = DAG([("a", "b"), ("b", "c")])
        self.assertListEqual(list(self.graph.predecessors("a")), [])
        self.assertListEqual(list(self.graph.predecessors("b")), ["a"])
        self.assertListEqual(list(self.graph.predecessors("c")), ["b"])

    def test_update_node_parents(self):
        self.graph.add_nodes_from(["a", "b", "c"])
        self.graph.add_edges_from([("a", "b"), ("b", "c")])
        self.assertListEqual(list(self.graph.predecessors("a")), [])
        self.assertListEqual(list(self.graph.predecessors("b")), ["a"])
        self.assertListEqual(list(self.graph.predecessors("c")), ["b"])

    def test_get_leaves(self):
        self.graph.add_edges_from(
            [("A", "B"), ("B", "C"), ("B", "D"), ("D", "E"), ("D", "F"), ("A", "G")]
        )
        self.assertEqual(sorted(self.graph.get_leaves()), sorted(["C", "G", "E", "F"]))

    def test_get_roots(self):
        self.graph.add_edges_from(
            [("A", "B"), ("B", "C"), ("B", "D"), ("D", "E"), ("D", "F"), ("A", "G")]
        )
        self.assertEqual(["A"], self.graph.get_roots())
        self.graph.add_edge("H", "G")
        self.assertEqual(sorted(["A", "H"]), sorted(self.graph.get_roots()))

    def test_init_with_cycle(self):
        self.assertRaises(ValueError, DAG, [("a", "a")])
        self.assertRaises(ValueError, DAG, [("a", "b"), ("b", "a")])
        self.assertRaises(ValueError, DAG, [("a", "b"), ("b", "c"), ("c", "a")])

    def test_get_ancestral_graph(self):
        dag = DAG([("A", "C"), ("B", "C"), ("D", "A"), ("D", "B")])
        anc_dag = dag.get_ancestral_graph(["A", "B"])
        self.assertEqual(set(anc_dag.edges()), set([("D", "A"), ("D", "B")]))
        self.assertRaises(ValueError, dag.get_ancestral_graph, ["A", "gibber"])

    def test_minimal_dseparator(self):
        # Without latent variables

        dag1 = DAG([("A", "B"), ("B", "C")])
        self.assertEqual(dag1.minimal_dseparator(start="A", end="C"), {"B"})

        dag2 = DAG([("A", "B"), ("B", "C"), ("C", "D"), ("A", "E"), ("E", "D")])
        self.assertEqual(dag2.minimal_dseparator(start="A", end="D"), {"C", "E"})

        dag3 = DAG(
            [("B", "A"), ("B", "C"), ("A", "D"), ("D", "C"), ("A", "E"), ("C", "E")]
        )
        self.assertEqual(dag3.minimal_dseparator(start="A", end="C"), {"B", "D"})

        # With latent variables

        dag_lat1 = DAG([("A", "B"), ("B", "C")], latents={"B"})
        self.assertIsNone(dag_lat1.minimal_dseparator(start="A", end="C"))

        dag_lat2 = DAG([("A", "D"), ("D", "B"), ("B", "C")], latents={"B"})
        self.assertEqual(dag_lat2.minimal_dseparator(start="A", end="C"), {"D"})

        dag_lat3 = DAG([("A", "B"), ("B", "D"), ("D", "C")], latents={"B"})
        self.assertEqual(dag_lat3.minimal_dseparator(start="A", end="C"), {"D"})

        dag_lat4 = DAG([("A", "B"), ("B", "C"), ("A", "D"), ("D", "C")], latents={"D"})
        self.assertIsNone(dag_lat4.minimal_dseparator(start="A", end="C"))

        dag_lat5 = DAG(
            [("A", "B"), ("B", "C"), ("A", "D"), ("D", "E"), ("E", "C")], latents={"E"}
        )
        self.assertEqual(dag_lat5.minimal_dseparator(start="A", end="C"), {"B", "D"})

    def test_to_daft(self):
        dag = DAG([("A", "C"), ("B", "C"), ("D", "A"), ("D", "B")])
        dag.to_daft(node_pos="circular")

        self.assertRaises(ValueError, dag.to_daft, node_pos="abcd")
        self.assertRaises(ValueError, dag.to_daft, node_pos={"A": (0, 2)})
        self.assertRaises(ValueError, dag.to_daft, node_pos=["random"])

        for layout in [
            "circular",
            "kamada_kawai",
            "planar",
            "random",
            "shell",
            "spring",
            #            "spectral", # TODO: Fails for latest networkx
            "spiral",
        ]:
            dag.to_daft(node_pos=layout)
            dag.to_daft(node_pos=layout, pgm_params={"observed_style": "inner"})
            dag.to_daft(
                node_pos=layout,
                edge_params={("A", "C"): {"label": 2}},
                node_params={"A": {"shape": "rectangle"}},
            )

    def test_random_dag(self):
        for i in range(10):
            n_nodes = 8
            edge_prob = np.random.uniform()
            dag = DAG.get_random(n_nodes=n_nodes, edge_prob=edge_prob)
            self.assertEqual(len(dag.nodes()), n_nodes)
            self.assertTrue(nx.is_directed_acyclic_graph(dag))
            self.assertTrue(len(dag.latents) == 0)

            node_names = [
                "a",
                "aa",
                "aaa",
                "aaaa",
                "aaaaa",
                "aaaaaa",
                "aaaaaaa",
                "aaaaaaaa",
            ]
            dag = DAG.get_random(
                n_nodes=n_nodes, edge_prob=edge_prob, node_names=node_names
            )
            self.assertEqual(len(dag.nodes()), n_nodes)
            self.assertEqual(sorted(dag.nodes()), node_names)
            self.assertTrue(nx.is_directed_acyclic_graph(dag))
            self.assertTrue(len(dag.latents) == 0)

        dag_latents = DAG.get_random(n_nodes=n_nodes, edge_prob=0.5, latents=True)

    def test_dag_fit(self):
        edge_list = [("A", "C"), ("B", "C")]
        for model in [DAG(edge_list), DiscreteBayesianNetwork(edge_list)]:
            data = pd.DataFrame(data={"A": [0, 0, 1], "B": [0, 1, 0], "C": [1, 1, 0]})
            pseudo_counts = {
                "A": [[9], [3]],
                "B": [[9], [3]],
                "C": [[9, 9, 9, 9], [3, 3, 3, 3]],
            }

            fitted_model_bayesian = model.fit(
                data,
                estimator=BayesianEstimator,
                prior_type="dirichlet",
                pseudo_counts=pseudo_counts,
            )
            self.assertEqual(
                fitted_model_bayesian.get_cpds("B"),
                TabularCPD("B", 2, [[11.0 / 15], [4.0 / 15]]),
            )

            fitted_model_mle = model.fit(data, estimator=MaximumLikelihoodEstimator)

            self.assertEqual(
                fitted_model_mle.get_cpds("B"),
                TabularCPD("B", 2, [[2.0 / 3], [1.0 / 3]]),
            )

            fitted_model_em = model.fit(data, estimator=ExpectationMaximization)

            self.assertEqual(
                fitted_model_em.get_cpds("B"),
                TabularCPD("B", 2, [[2.0 / 3], [1.0 / 3]]),
            )

    def tearDown(self):
        del self.graph

    def test_edge_strength_basic(self):
        """Test basic functionality and numerical values"""
        dag = DAG([("X", "Y"), ("Z", "Y")])
        data = pd.DataFrame({"X": [0, 1, 0, 1], "Y": [1, 3, 0, 2], "Z": [1, 1, 0, 0]})
        strengths = dag.edge_strength(data)

        # Test return type and structure
        self.assertTrue(isinstance(strengths, dict))
        self.assertEqual(set(strengths.keys()), {("X", "Y"), ("Z", "Y")})
        self.assertTrue(all(isinstance(v, float) for v in strengths.values()))

        # Test numerical values from docstring example
        self.assertAlmostEqual(strengths[("X", "Y")], 0.999999754781587, places=15)
        self.assertAlmostEqual(strengths[("Z", "Y")], 0.9999989719376587, places=15)

    def test_edge_strength_specific_edge(self):
        """Test computing strength for specific edge"""
        dag = DAG([("X", "Y"), ("Z", "Y")])
        data = pd.DataFrame({"X": [0, 1, 0, 1], "Y": [1, 3, 0, 2], "Z": [1, 1, 0, 0]})
        strength_xy = dag.edge_strength(data, edges=("X", "Y"))

        # Test structure
        self.assertEqual(set(strength_xy.keys()), {("X", "Y")})

        # Test numerical value from docstring
        self.assertAlmostEqual(strength_xy[("X", "Y")], 0.999999754781587, places=15)

    def test_edge_strength_empty_dag(self):
        """Test edge case with empty DAG"""
        empty_dag = DAG()
        data = pd.DataFrame({"X": [0, 1, 0, 1], "Y": [1, 3, 0, 2], "Z": [1, 1, 0, 0]})
        self.assertEqual(empty_dag.edge_strength(data), {})

    def test_edge_strength_multiple_edges(self):
        """Test computing strength for multiple specific edges"""
        dag = DAG([("X", "Y"), ("Z", "Y")])
        data = pd.DataFrame({"X": [0, 1, 0, 1], "Y": [1, 3, 0, 2], "Z": [1, 1, 0, 0]})
        strengths = dag.edge_strength(data, edges=[("X", "Y"), ("Z", "Y")])

        # Test structure
        self.assertEqual(set(strengths.keys()), {("X", "Y"), ("Z", "Y")})

        # Test numerical values from docstring
        self.assertAlmostEqual(strengths[("X", "Y")], 0.999999754781587, places=15)
        self.assertAlmostEqual(strengths[("Z", "Y")], 0.9999989719376587, places=15)

class TestDAGParser(unittest.TestCase):
    def test_from_lavaan(self):
        model_str = """# %load model.lav
                       # measurement model
                         ind60 =~ x1 + x2 + x3
                         dem60 =~ y1 + y2 + y3 + y4
                         dem65 =~ y5 + y6 + y7 + y8
                       # regressions
                         dem60 ~ ind60
                         dem65 ~ ind60 + dem60
                       """
        model_from_str = DAG.from_lavaan(string=model_str)

        with open("test_model.lav", "w") as f:
            f.write(model_str)
        model_from_file = DAG.from_lavaan(filename="test_model.lav")
        os.remove("test_model.lav")

        expected_edges = set(
            [
                ("ind60", "x1"),
                ("ind60", "x2"),
                ("ind60", "x3"),
                ("ind60", "dem60"),
                ("ind60", "dem65"),
                ("dem60", "dem65"),
                ("dem60", "y1"),
                ("dem60", "y2"),
                ("dem60", "y3"),
                ("dem60", "y4"),
                ("dem65", "y5"),
                ("dem65", "y6"),
                ("dem65", "y7"),
                ("dem65", "y8"),
            ]
        )

        expected_latents = set(["dem60", "dem65", "ind60"])
        self.assertEqual(set(model_from_str.edges()), expected_edges)
        self.assertEqual(set(model_from_file.edges()), expected_edges)
        self.assertEqual(set(model_from_str.latents), expected_latents)
        self.assertEqual(set(model_from_file.latents), expected_latents)

    def test_from_lavaan_with_residual_correlation(self):
        model_str = """# %load model_with_residual_correlation.lav
                       # measurement model
                         ind60 =~ x1 + x2 + x3
                       # regressions
                         dem60 ~ ind60
                       # residual correlations
                         y1 ~~ y5
                       """

        model_from_str = DAG.from_lavaan(string=model_str)
        expected_edges = set(
            [
                ("ind60", "x1"),
                ("ind60", "x2"),
                ("ind60", "x3"),
                ("ind60", "dem60"),
            ]
        )

        expected_latents = set(["ind60"])
        self.assertEqual(set(model_from_str.edges()), expected_edges)
        self.assertEqual(set(model_from_str.latents), expected_latents)

    def test_from_dagitty(self):
        model_str = """
            dag{
                smoking "carry matches" [e] ; cancer [o]
                smoking -> {"carry matches" -> cancer} smoking <-> coffee
            }"""
        model_from_str = DAG.from_dagitty(string=model_str)

        with open("test_model.dagitty", "w") as f:
            f.write(model_str)
        model_from_file = DAG.from_dagitty(filename="test_model.dagitty")
        os.remove("test_model.dagitty")

        expected_edges = set(
            [
                ("smoking", "cancer"),
                ("smoking", "carry matches"),
                ("carry matches", "cancer"),
                ("u_coffee_smoking", "coffee"),
                ("u_coffee_smoking", "smoking"),
            ]
        )

        expected_latents = set(["u_coffee_smoking"])
        self.assertEqual(set(model_from_str.edges()), expected_edges)
        self.assertEqual(set(model_from_file.edges()), expected_edges)
        self.assertEqual(set(model_from_str.latents), expected_latents)
        self.assertEqual(set(model_from_file.latents), expected_latents)

    def test_from_daggitty_single_line_with_group_of_vars(self):
        dag = DAG.from_dagitty(
            'dag{ bb="0,0,1,1" X [l, pos="-1.228,-1.145"] X-> {Y Z}  Z ->A ->B <- C}'
        )
        self.assertEqual(
            set(dag.edges()),
            set([("X", "Z"), ("X", "Y"), ("Z", "A"), ("A", "B"), ("C", "B")]),
        )
        self.assertEqual(set(dag.latents), set(["X"]))

    def test_from_dagitty_multiline_with_display_info(self):
        dag = DAG.from_dagitty(
            """
                dag {
                bb="-1.728,-4.67,2.587,4.156"
                123 [pos="2.087,3.420"]
                X.1 [pos="-1.228,-1.145"]
                Y [pos="-0.725,-3.934"]
                Z [latent, pos="-0.135,1.659"]
                X.1 -> Y [pos="-0.300,-0.082"]
                X.1 -> Z
                Z -> 123
                }
        """
        )
        self.assertEqual(
            set(dag.edges()), set([("X.1", "Y"), ("X.1", "Z"), ("Z", "123")])
        )
        self.assertEqual(set(dag.latents), set(["Z"]))

class TestDAGMoralization(unittest.TestCase):
    def setUp(self):
        self.graph = DAG()
        self.graph.add_edges_from([("diff", "grade"), ("intel", "grade")])

    def test_get_parents(self):
        self.assertListEqual(sorted(self.graph.get_parents("grade")), ["diff", "intel"])

    def test_moralize(self):
        moral_graph = self.graph.moralize()
        self.assertListEqual(
            hf.recursive_sorted(moral_graph.edges()),
            [["diff", "grade"], ["diff", "intel"], ["grade", "intel"]],
        )

    def test_moralize_disconnected(self):
        graph_copy = self.graph.copy()
        graph_copy.add_node("disconnected")
        moral_graph = graph_copy.moralize()
        self.assertListEqual(
            hf.recursive_sorted(moral_graph.edges()),
            [["diff", "grade"], ["diff", "intel"], ["grade", "intel"]],
        )
        self.assertEqual(
            sorted(moral_graph.nodes()), ["diff", "disconnected", "grade", "intel"]
        )

    def test_get_children(self):
        self.assertListEqual(sorted(self.graph.get_children("diff")), ["grade"])

    def tearDown(self):
        del self.graph

class TestDoOperator(unittest.TestCase):
    def setUp(self):
        self.g1 = DAG([("X", "A"), ("A", "Y"), ("A", "B")])
        self.g2 = DAG([("X", "A"), ("A", "Y"), ("A", "B"), ("Z", "Y")])

    def test_do(self):
        dag_do_x = self.g1.do("A")
        self.assertEqual(set(dag_do_x.nodes()), set(self.g1.nodes()))
        self.assertEqual(sorted(list(dag_do_x.edges())), [("A", "B"), ("A", "Y")])

        dag_do_x = self.g2.do(["A", "Y"])
        self.assertEqual(set(dag_do_x.nodes()), set(self.g2.nodes()))
        self.assertEqual(sorted(list(dag_do_x.edges())), [("A", "B")])

class TestPDAG(unittest.TestCase):
    def setUp(self):
        self.pdag_mix = PDAG(
            directed_ebunch=[("A", "C"), ("D", "C")],
            undirected_ebunch=[("B", "A"), ("B", "D")],
        )
        self.pdag_dir = PDAG(
            directed_ebunch=[("A", "B"), ("D", "B"), ("A", "C"), ("D", "C")]
        )
        self.pdag_undir = PDAG(
            undirected_ebunch=[("A", "C"), ("D", "C"), ("B", "A"), ("B", "D")]
        )
        self.pdag_latent = PDAG(
            directed_ebunch=[("A", "C"), ("D", "C")],
            undirected_ebunch=[("B", "A"), ("B", "D")],
            latents=["A", "D"],
        )

    def test_init_normal(self):
        # Mix directed and undirected
        directed_edges = [("A", "C"), ("D", "C")]
        undirected_edges = [("B", "A"), ("B", "D")]
        pdag = PDAG(directed_ebunch=directed_edges, undirected_ebunch=undirected_edges)
        expected_edges = {
            ("A", "C"),
            ("D", "C"),
            ("A", "B"),
            ("B", "A"),
            ("B", "D"),
            ("D", "B"),
        }
        self.assertEqual(set(pdag.edges()), expected_edges)
        self.assertEqual(set(pdag.nodes()), {"A", "B", "C", "D"})
        self.assertEqual(pdag.directed_edges, set(directed_edges))
        self.assertEqual(pdag.undirected_edges, set(undirected_edges))

        directed_edges = [("A", "C"), ("D", "C")]
        undirected_edges = [("B", "A"), ("B", "D")]
        pdag = PDAG(
            directed_ebunch=directed_edges,
            undirected_ebunch=undirected_edges,
            latents=["A", "C"],
        )
        expected_edges = {
            ("A", "C"),
            ("D", "C"),
            ("A", "B"),
            ("B", "A"),
            ("B", "D"),
            ("D", "B"),
        }
        self.assertEqual(set(pdag.edges()), expected_edges)
        self.assertEqual(set(pdag.nodes()), {"A", "B", "C", "D"})
        self.assertEqual(pdag.directed_edges, set(directed_edges))
        self.assertEqual(pdag.undirected_edges, set(undirected_edges))
        self.assertEqual(pdag.latents, set(["A", "C"]))

        # Only undirected
        undirected_edges = [("A", "C"), ("D", "C"), ("B", "A"), ("B", "D")]
        pdag = PDAG(undirected_ebunch=undirected_edges)
        expected_edges = {
            ("A", "C"),
            ("C", "A"),
            ("D", "C"),
            ("C", "D"),
            ("B", "A"),
            ("A", "B"),
            ("B", "D"),
            ("D", "B"),
        }
        self.assertEqual(set(pdag.edges()), expected_edges)
        self.assertEqual(set(pdag.nodes()), {"A", "B", "C", "D"})
        self.assertEqual(pdag.directed_edges, set())
        self.assertEqual(pdag.undirected_edges, set(undirected_edges))

        undirected_edges = [("A", "C"), ("D", "C"), ("B", "A"), ("B", "D")]
        pdag = PDAG(undirected_ebunch=undirected_edges, latents=["A", "D"])
        expected_edges = {
            ("A", "C"),
            ("C", "A"),
            ("D", "C"),
            ("C", "D"),
            ("B", "A"),
            ("A", "B"),
            ("B", "D"),
            ("D", "B"),
        }
        self.assertEqual(set(pdag.edges()), expected_edges)
        self.assertEqual(set(pdag.nodes()), {"A", "B", "C", "D"})
        self.assertEqual(pdag.directed_edges, set())
        self.assertEqual(pdag.undirected_edges, set(undirected_edges))
        self.assertEqual(pdag.latents, set(["A", "D"]))

        # Only directed
        directed_edges = [("A", "B"), ("D", "B"), ("A", "C"), ("D", "C")]
        pdag = PDAG(directed_ebunch=directed_edges)
        self.assertEqual(set(pdag.edges()), set(directed_edges))
        self.assertEqual(set(pdag.nodes()), {"A", "B", "C", "D"})
        self.assertEqual(pdag.directed_edges, set(directed_edges))
        self.assertEqual(pdag.undirected_edges, set())

        directed_edges = [("A", "B"), ("D", "B"), ("A", "C"), ("D", "C")]
        pdag = PDAG(directed_ebunch=directed_edges, latents=["D"])
        self.assertEqual(set(pdag.edges()), set(directed_edges))
        self.assertEqual(set(pdag.nodes()), {"A", "B", "C", "D"})
        self.assertEqual(pdag.directed_edges, set(directed_edges))
        self.assertEqual(pdag.undirected_edges, set())
        self.assertEqual(pdag.latents, set(["D"]))

    def test_all_neighrors(self):
        directed_edges = [("A", "C"), ("D", "C")]
        undirected_edges = [("B", "A"), ("B", "D")]
        pdag = PDAG(directed_ebunch=directed_edges, undirected_ebunch=undirected_edges)

        self.assertEqual(pdag.all_neighbors(node="A"), {"B", "C"})
        self.assertEqual(pdag.all_neighbors(node="B"), {"A", "D"})
        self.assertEqual(pdag.all_neighbors(node="C"), {"A", "D"})
        self.assertEqual(pdag.all_neighbors(node="D"), {"B", "C"})

    def test_directed_children(self):
        directed_edges = [("A", "C"), ("D", "C")]
        undirected_edges = [("B", "A"), ("B", "D")]
        pdag = PDAG(directed_ebunch=directed_edges, undirected_ebunch=undirected_edges)

        self.assertEqual(pdag.directed_children(node="A"), {"C"})
        self.assertEqual(pdag.directed_children(node="B"), set())
        self.assertEqual(pdag.directed_children(node="C"), set())

    def test_directed_parents(self):
        directed_edges = [("A", "C"), ("D", "C")]
        undirected_edges = [("B", "A"), ("B", "D")]
        pdag = PDAG(directed_ebunch=directed_edges, undirected_ebunch=undirected_edges)

        self.assertEqual(pdag.directed_parents(node="A"), set())
        self.assertEqual(pdag.directed_parents(node="B"), set())
        self.assertEqual(pdag.directed_parents(node="C"), {"A", "D"})

    def test_has_directed_edge(self):
        directed_edges = [("A", "C"), ("D", "C")]
        undirected_edges = [("B", "A"), ("B", "D")]
        pdag = PDAG(directed_ebunch=directed_edges, undirected_ebunch=undirected_edges)

        self.assertTrue(pdag.has_directed_edge("A", "C"))
        self.assertTrue(pdag.has_directed_edge("D", "C"))
        self.assertFalse(pdag.has_directed_edge("A", "B"))
        self.assertFalse(pdag.has_directed_edge("B", "A"))

    def test_has_undirected_edge(self):
        directed_edges = [("A", "C"), ("D", "C")]
        undirected_edges = [("B", "A"), ("B", "D")]
        pdag = PDAG(directed_ebunch=directed_edges, undirected_ebunch=undirected_edges)

        self.assertFalse(pdag.has_undirected_edge("A", "C"))
        self.assertFalse(pdag.has_undirected_edge("D", "C"))
        self.assertTrue(pdag.has_undirected_edge("A", "B"))
        self.assertTrue(pdag.has_undirected_edge("B", "A"))

    def test_undirected_neighbors(self):
        directed_edges = [("A", "C"), ("D", "C")]
        undirected_edges = [("B", "A"), ("B", "D")]
        pdag = PDAG(directed_ebunch=directed_edges, undirected_ebunch=undirected_edges)

        self.assertEqual(pdag.undirected_neighbors(node="A"), {"B"})
        self.assertEqual(pdag.undirected_neighbors(node="B"), {"A", "D"})
        self.assertEqual(pdag.undirected_neighbors(node="C"), set())
        self.assertEqual(pdag.undirected_neighbors(node="D"), {"B"})

    def test_orient_undirected_edge(self):
        directed_edges = [("A", "C"), ("D", "C")]
        undirected_edges = [("B", "A"), ("B", "D")]
        pdag = PDAG(directed_ebunch=directed_edges, undirected_ebunch=undirected_edges)

        mod_pdag = pdag.orient_undirected_edge("B", "A", inplace=False)
        self.assertEqual(
            set(mod_pdag.edges()),
            {("A", "C"), ("D", "C"), ("B", "A"), ("B", "D"), ("D", "B")},
        )
        self.assertEqual(mod_pdag.undirected_edges, {("B", "D")})
        self.assertEqual(mod_pdag.directed_edges, {("A", "C"), ("D", "C"), ("B", "A")})

        pdag.orient_undirected_edge("B", "A", inplace=True)
        self.assertEqual(
            set(pdag.edges()),
            {("A", "C"), ("D", "C"), ("B", "A"), ("B", "D"), ("D", "B")},
        )
        self.assertEqual(pdag.undirected_edges, {("B", "D")})
        self.assertEqual(pdag.directed_edges, {("A", "C"), ("D", "C"), ("B", "A")})

        self.assertRaises(
            ValueError, pdag.orient_undirected_edge, "B", "A", inplace=True
        )

    def test_copy(self):
        pdag_copy = self.pdag_mix.copy()
        expected_edges = {
            ("A", "C"),
            ("D", "C"),
            ("A", "B"),
            ("B", "A"),
            ("B", "D"),
            ("D", "B"),
        }
        expected_dir = [("A", "C"), ("D", "C")]
        expected_undir = [("B", "A"), ("B", "D")]
        self.assertEqual(set(pdag_copy.edges()), expected_edges)
        self.assertEqual(set(pdag_copy.nodes()), {"A", "B", "C", "D"})
        self.assertEqual(pdag_copy.directed_edges, set([("A", "C"), ("D", "C")]))
        self.assertEqual(pdag_copy.undirected_edges, set([("B", "A"), ("B", "D")]))
        self.assertEqual(pdag_copy.latents, set())

        pdag_copy = self.pdag_latent.copy()
        expected_edges = {
            ("A", "C"),
            ("D", "C"),
            ("A", "B"),
            ("B", "A"),
            ("B", "D"),
            ("D", "B"),
        }
        expected_dir = [("A", "C"), ("D", "C")]
        expected_undir = [("B", "A"), ("B", "D")]
        self.assertEqual(set(pdag_copy.edges()), expected_edges)
        self.assertEqual(set(pdag_copy.nodes()), {"A", "B", "C", "D"})
        self.assertEqual(pdag_copy.directed_edges, set([("A", "C"), ("D", "C")]))
        self.assertEqual(pdag_copy.undirected_edges, set([("B", "A"), ("B", "D")]))
        self.assertEqual(pdag_copy.latents, set(["A", "D"]))

    def test_pdag_to_dag(self):
        # PDAG no: 1  Possibility of creating a v-structure
        pdag = PDAG(
            directed_ebunch=[("A", "B"), ("C", "B")],
            undirected_ebunch=[("C", "D"), ("D", "A")],
        )
        dag = pdag.to_dag()
        self.assertTrue(("A", "B") in dag.edges())
        self.assertTrue(("C", "B") in dag.edges())
        self.assertFalse((("A", "D") in dag.edges()) and (("C", "D") in dag.edges()))
        self.assertTrue(len(dag.edges()) == 4)

        # With latents
        pdag = PDAG(
            directed_ebunch=[("A", "B"), ("C", "B")],
            undirected_ebunch=[("C", "D"), ("D", "A")],
            latents=["A"],
        )
        dag = pdag.to_dag()
        self.assertTrue(("A", "B") in dag.edges())
        self.assertTrue(("C", "B") in dag.edges())
        self.assertFalse((("A", "D") in dag.edges()) and (("C", "D") in dag.edges()))
        self.assertEqual(dag.latents, set(["A"]))
        self.assertTrue(len(dag.edges()) == 4)

        # PDAG no: 2  No possibility of creation of v-structure.
        pdag = PDAG(
            directed_ebunch=[("B", "C"), ("A", "C")], undirected_ebunch=[("A", "D")]
        )
        dag = pdag.to_dag()
        self.assertTrue(("B", "C") in dag.edges())
        self.assertTrue(("A", "C") in dag.edges())
        self.assertTrue((("A", "D") in dag.edges()) or (("D", "A") in dag.edges()))

        # With latents
        pdag = PDAG(
            directed_ebunch=[("B", "C"), ("A", "C")],
            undirected_ebunch=[("A", "D")],
            latents=["A"],
        )
        dag = pdag.to_dag()
        self.assertTrue(("B", "C") in dag.edges())
        self.assertTrue(("A", "C") in dag.edges())
        self.assertTrue((("A", "D") in dag.edges()) or (("D", "A") in dag.edges()))
        self.assertEqual(dag.latents, set(["A"]))

        # PDAG no: 3  Already existing v-structure, possibility to add another
        pdag = PDAG(
            directed_ebunch=[("B", "C"), ("A", "C")], undirected_ebunch=[("C", "D")]
        )
        dag = pdag.to_dag()
        expected_edges = {("B", "C"), ("C", "D"), ("A", "C")}
        self.assertEqual(expected_edges, set(dag.edges()))

        # With latents
        pdag = PDAG(
            directed_ebunch=[("B", "C"), ("A", "C")],
            undirected_ebunch=[("C", "D")],
            latents=["A"],
        )
        dag = pdag.to_dag()
        expected_edges = {("B", "C"), ("C", "D"), ("A", "C")}
        self.assertEqual(expected_edges, set(dag.edges()))
        self.assertEqual(dag.latents, set(["A"]))

    def test_pdag_to_cpdag(self):
        pdag = PDAG(directed_ebunch=[("A", "B")], undirected_ebunch=[("B", "C")])
        cpdag = pdag.apply_meeks_rules(apply_r4=True)
        self.assertSetEqual(set(cpdag.edges()), {("A", "B"), ("B", "C")})

        pdag = PDAG(
            directed_ebunch=[("A", "B")], undirected_ebunch=[("B", "C"), ("C", "D")]
        )
        cpdag = pdag.apply_meeks_rules(apply_r4=True)
        self.assertSetEqual(set(cpdag.edges()), {("A", "B"), ("B", "C"), ("C", "D")})

        pdag = PDAG(
            directed_ebunch=[("A", "B"), ("D", "C")], undirected_ebunch=[("B", "C")]
        )
        cpdag = pdag.apply_meeks_rules(apply_r4=True)
        self.assertSetEqual(
            set(cpdag.edges()), {("A", "B"), ("D", "C"), ("B", "C"), ("C", "B")}
        )

        pdag = PDAG(
            directed_ebunch=[("A", "B"), ("D", "C"), ("D", "B")],
            undirected_ebunch=[("B", "C")],
        )
        cpdag = pdag.apply_meeks_rules(apply_r4=True)
        self.assertSetEqual(
            set(cpdag.edges()), {("A", "B"), ("D", "C"), ("D", "B"), ("B", "C")}
        )

        pdag = PDAG(
            directed_ebunch=[("A", "B"), ("B", "C")], undirected_ebunch=[("A", "C")]
        )
        cpdag = pdag.apply_meeks_rules(apply_r4=True)
        self.assertSetEqual(set(cpdag.edges()), {("A", "B"), ("B", "C"), ("A", "C")})

        pdag = PDAG(
            directed_ebunch=[("A", "B"), ("B", "C"), ("D", "C")],
            undirected_ebunch=[("A", "C")],
        )
        cpdag = pdag.apply_meeks_rules(apply_r4=True)
        self.assertSetEqual(
            set(cpdag.edges()), {("A", "B"), ("B", "C"), ("A", "C"), ("D", "C")}
        )

        # Examples taken from Perkovi\`c 2017.
        pdag = PDAG(
            directed_ebunch=[("V1", "X")],
            undirected_ebunch=[("X", "V2"), ("V2", "Y"), ("X", "Y")],
        )
        cpdag = pdag.apply_meeks_rules(apply_r4=True)
        self.assertEqual(
            set(cpdag.edges()),
            {("V1", "X"), ("X", "V2"), ("X", "Y"), ("V2", "Y"), ("Y", "V2")},
        )

        pdag = PDAG(
            directed_ebunch=[("Y", "X")],
            undirected_ebunch=[("V1", "X"), ("X", "V2"), ("V2", "Y")],
        )
        cpdag = pdag.apply_meeks_rules(apply_r4=True)
        self.assertEqual(
            set(cpdag.edges()),
            {
                ("X", "V1"),
                ("Y", "X"),
                ("X", "V2"),
                ("V2", "X"),
                ("V2", "Y"),
                ("Y", "V2"),
            },
        )

        # Examples from Bang 2024
        pdag = PDAG(
            directed_ebunch=[("B", "D"), ("C", "D")],
            undirected_ebunch=[("A", "D"), ("A", "C")],
        )
        cpdag = pdag.apply_meeks_rules(apply_r4=True, debug=True)
        self.assertEqual(
            set(cpdag.edges()), {("B", "D"), ("D", "A"), ("C", "A"), ("C", "D")}
        )

        pdag = PDAG(
            directed_ebunch=[("A", "B"), ("C", "B")],
            undirected_ebunch=[("D", "B"), ("D", "A"), ("D", "C")],
        )
        cpdag = pdag.apply_meeks_rules(apply_r4=True)
        self.assertSetEqual(
            set(cpdag.edges()),
            {
                ("A", "B"),
                ("C", "B"),
                ("D", "B"),
                ("D", "A"),
                ("A", "D"),
                ("D", "C"),
                ("C", "D"),
            },
        )

        undirected_edges = [("A", "C"), ("B", "C"), ("D", "C")]
        directed_edges = [("B", "D"), ("D", "A")]

        pdag = PDAG(directed_ebunch=directed_edges, undirected_ebunch=undirected_edges)
        mpdag = pdag.apply_meeks_rules(apply_r4=True)
        self.assertSetEqual(
            set(mpdag.edges()),
            set(
                [
                    ("C", "A"),
                    ("C", "B"),
                    ("B", "C"),
                    ("B", "D"),
                    ("D", "A"),
                    ("D", "C"),
                    ("C", "D"),
                ]
            ),
        )

        pdag = PDAG(directed_ebunch=directed_edges, undirected_ebunch=undirected_edges)
        pdag = pdag.apply_meeks_rules()
        self.assertSetEqual(
            set(pdag.edges()),
            set(
                [
                    ("A", "C"),
                    ("C", "A"),
                    ("C", "B"),
                    ("B", "C"),
                    ("B", "D"),
                    ("D", "A"),
                    ("D", "C"),
                    ("C", "D"),
                ]
            ),
        )

        pdag_inp = PDAG(
            directed_ebunch=directed_edges, undirected_ebunch=undirected_edges
        )
        pdag_inp.apply_meeks_rules(inplace=True)
        self.assertSetEqual(
            set(pdag_inp.edges()),
            set(
                [
                    ("A", "C"),
                    ("C", "A"),
                    ("C", "B"),
                    ("B", "C"),
                    ("B", "D"),
                    ("D", "A"),
                    ("D", "C"),
                    ("C", "D"),
                ]
            ),
        )

Manas-7854 committed:
Merge branch 'dev' of https://github.com/Manas-7854/pgmpy into dev

ankurankan commented on pgmpy/base/DAG.py:
```diff
@@ -1257,6 +1257,107 @@ def copy(self):
+        elif isinstance(edges, tuple) and len(edges) == 2:
+            edges_to_compute = [edges]
+        # If edges is a list of edge tuples
+        else:
```
I would suggest checking there that the argument here is a list of 2-tuples. And add another else, which throws an ValueError that the input by the user is not correct with some information on what we expect.

ankurankan commented on pgmpy/base/DAG.py:
```diff
@@ -1257,6 +1257,107 @@ def copy(self):
+            if (
+                x in self.latents
+                or y in self.latents
+                or any(parent in self.latents for parent in pa_X + pa_Y)
```
This would need to be just pa_Y. `pa_X` is not needed.

ankurankan commented on pgmpy/base/DAG.py:
```diff
@@ -1257,6 +1257,107 @@ def copy(self):
+            x, y = edge
+
+            # Get parents of x and y using get_parents instead of predecessors
+            pa_X = self.get_parents(x)
```
We don't need pa_X.

ankurankan commented on pgmpy/base/DAG.py:
```diff
@@ -1257,6 +1257,107 @@ def copy(self):
+            )
+
+            # Store the edge strength
+            strengths[edge] = result[0]
```
I think we should also store the computed strengths in the graph itself. This could help us with plotting functionality. For each edge, we can do something like:
```
self.edges[edge]['strength'] = compute_value
```

ankurankan commented on pgmpy/base/DAG.py:
```diff
@@ -1257,6 +1257,107 @@ def copy(self):
+            pa_Y = self.get_parents(y)
+
+            # Check if either x or y is a latent node
+            if (
```
In the case when we are computing the strength for a list of edges, it might be better to show a warning for the edge that we can't compute the strength for and compute for the remaining. What do you think?

Manas-7854 replied:
yeah that would be better, will do that

ankurankan commented on pgmpy/tests/test_base/test_DAG.py:
```diff
@@ -360,6 +361,52 @@ def test_dag_fit(self):
+
+        # Test numerical values from docstring
+        self.assertAlmostEqual(strengths[("X", "Y")], 0.999999754781587, places=15)
+        self.assertAlmostEqual(strengths[("Z", "Y")], 0.9999989719376587, places=15)
```
Did the LinearGaussian BN tests still didn't work when comparing to the square root of strength?

ankurankan commented on pgmpy/tests/test_base/test_DAG.py:
```diff
@@ -360,6 +361,52 @@ def test_dag_fit(self):
     def tearDown(self):
         del self.graph
 
+    def test_edge_strength_basic(self):
```
Could you also add some tests with latent variables?

Manas-7854 committed:
Merge branch 'pgmpy:dev' into dev

Manas-7854 committed:
Merge remote-tracking branch 'refs/remotes/origin/dev' into dev

Manas-7854 committed:
test for latent,store strength in graph,minor fixe

Manas-7854 committed:
minor changes, latent test, LGBN test

Manas-7854 committed:
updated docstring with lgbn

ankurankan committed:
Minor modifications to the docstring and code

ankurankan reviewed (approved):
@Manas-7854 Looks good to me. I have added a few minor changes in docstring and code.
